### DISCUSSION TITLE
Unmarshalling LCD API response into Protobuf types fails

### DISCUSSION THREAD

**Discussion by freak12techno** [cosmos/community#69] [Tooling] [answered]
I need to fetch the validators list of a cosmos-sdk chain from LCD API of the node into the Protobuf types (I need specifically LCD API node, not RPC or gRPC node, these have no issues) or somehow convert it to these Protobuf types.

I try to query the chain with a code that's something like this:
```
import (
    stakingTypes "github.com/cosmos/cosmos-sdk/x/staking/types"
)

func (f *CosmosLcdDataFetcher) GetValidators() (*types.ChainValidators, error) {
    var validatorsResponse stakingTypes.QueryValidatorsResponse
    err := f.Client.Get(
        "/cosmos/staking/v1beta1/validators?status=BOND_STATUS_BONDED&pagination.limit=1000",
        &validatorsResponse,
    )
    if err != nil {
        return nil, err
    }
}
```

(the `f.Client` is a wrapper for HTTP client that fetches the URL relative to the host I need and unmarshals its JSON response into the second argument of `.Get()` function call)

This yields the following error:
```
json: cannot unmarshal string into Go struct field Validator.validators.status of type types.BondStatus
```

I know I can write my own JSON types for the API response and unmarshal into it and then convert manually, but this isn't as convenient + I need to use some of the protobuf types' internal methods (such as GetConsAddr() or GetMoniker()).

Wonder if there's something built-in in cosmos-sdk that I can use as a client to LCD node that will return the type I need, or some other way to fetch the data and convert it into Protobuf types?

**Comment by tac0turtle** [comment]
LCD doesnt exist anymore. it is either grpc or rest. Which one are you trying to use here?

**Reply by freak12techno** [reply, marked as answer]
REST. I already managed to fix it btw, only needed to parse it via legacy amino, so feel free to mark it as resolved:

```
import (
    "github.com/cosmos/cosmos-sdk/codec"
    codecTypes "github.com/cosmos/cosmos-sdk/codec/types"
    "github.com/cosmos/cosmos-sdk/std"
    stakingTypes "github.com/cosmos/cosmos-sdk/x/staking/types"
)

    interfaceRegistry := codecTypes.NewInterfaceRegistry()
    std.RegisterInterfaces(interfaceRegistry)
    parseCodec := codec.NewProtoCodec(interfaceRegistry)

    // function that does a HTTP query and returns JSON response as []byte 
    bytes, err := client.GetPlain(
        "/cosmos/staking/v1beta1/validators?status=BOND_STATUS_BONDED&pagination.limit=1000",
    )

    if err != nil {
        return nil, err
    }

    var validatorsResponse stakingTypes.QueryValidatorsResponse

    if err := parseCodec.UnmarshalJSON(bytes, &validatorsResponse); err != nil {
        return nil, err
    }

    // everything's ok, you can work with validatorsResponse same way as you'd do
    // if you query it via /abci_query or gRPC
```
